components/atom/checkbox/src/index.js

```diff
@@ -1,41 +1,61 @@
-import React from 'react'
+import React, {useRef} from 'react'
 import PropTypes from 'prop-types'
 import cx from 'classnames'
 
 const BASE_CLASS = 'sui-AtomCheckbox'
 
 const AtomCheckbox = ({
-  id,
-  name,
-  disabled,
   checked = false,
-  intermediate = false,
   checkedIcon: CheckedIcon,
+  disabled,
+  id,
+  intermediate = false,
   intermediateIcon: IntermediateIcon,
+  isNative: isNativeProp = false,
+  name,
   onChange: onChangeFromProps = () => {},
   ...props
 }) => {
+  const inputRef = useRef()
+  const hasNotCustomIcons = !CheckedIcon && !IntermediateIcon
+  const isNative = isNativeProp || hasNotCustomIcons
+  const isIntermediate = intermediate && !checked
+
+  const updateNativeIndeterminate = () => {
+    inputRef.current && (inputRef.current.indeterminate = isIntermediate)
+  }
+
   const handleChange = ev => {
+    // Handler doesn't necessarily trigger render, but browser could still set
+    // native indeterminate property which may end up in a mismatch between it
+    // and the component's prop, so native value should be kept updated here.
+    updateNativeIndeterminate()
+
     const {checked, name} = ev.target
     if (!disabled) onChangeFromProps(ev, {name, value: checked})
   }
 
   const className = cx(BASE_CLASS, {
     'is-checked': checked,
-    'is-disabled': disabled
+    'is-disabled': disabled,
+    [`${BASE_CLASS}--native`]: isNative
   })
 
+  // Keep native indeterminate property updated every render
+  updateNativeIndeterminate()
+
   return (
     <label className={className}>
-      {checked && <CheckedIcon />}
-      {intermediate && !checked && <IntermediateIcon />}
+      {!isNative && checked && <CheckedIcon />}
```

Comment:
just as a suggestion if using `isIntermediate`  it would be nicer to use `isChecked`

Reply:
I totally agree the `is` prefix is cooler 👍

Unfortunately the currently exposed props were already named `checked` and `intermediate`, which cannot be changed without releasing a major,  while `isIntermediate` is just a new inner constant that is used to disable intermediate behaviour when `checked` prop is active (there may be better names for it 🤣). This makes `checked` prop to become the absolute priority if set along with `intermediate` prop (e.g. by accident).

We could rename outer `checked` prop to `isChecked` just for inner usage, but I think it's convenient to avoid that kind of [mental mapping](https://github.com/ryanmcdermott/clean-code-javascript#avoid-mental-mapping) and keep original prop names the same when possible.